lib/jwt/jwk/key_base.rb

```diff
@@ -15,14 +15,19 @@ def initialize(options)
           options = { kid: options }
         end
 
-        @kid           = options[:kid]
+        @common_parameters = options[:common_parameters] || {}
```

Comment:
Can we use some other word than `common`, feels to generic :). I like the word `parameter` because it's what is used in the RFC.

My suggestion would be`jwk_parameters`

Reply:
I had scanned the RFC for some name to use and these kinds of parameters were only ever termed `common parameters` or `key parameters that are not algorithm specific`.
I agree that `common_parameters` may be too generic. `jwk_parameters` or `parameters` on the other hand do not seem to be restrictive enough for parameters. E.g. in an RSA key I would intuitively expect to be able to set `n` inside the `jwk_parameters`.

Here are some more ideas, none of which are really great IMO:
- `auxiliary_parameters` or `aux_params`
- `additional_parameters`
- `descriptive_parameters` or `desc_params`

Reply:
This is kinda invalidating the initial idea, but what if we would not have these parameters as "exta" but move the importing into the object itself?

Some pseudo usage example of what I was thinking:
```
jwk_data = { kty:'RSA', kid:'thekid', n: 'xx' e: 'xx' d: 'xx' p: 'xx' q: 'xx' dp: 'xx' dq: 'xx' qi: 'xx', use:'sig' }

jwk = ::JWT::JWK.new(jwk_data) # still alias to create_from

jwk.is_a?(::JWT::JWK::RSA)
jwk.keypair # Would generate the RSA key from the parameters or keypair would be generated on initialization
jwk.export(include_private) == jwk_data 
jwk[:use] == 'sig'
jwk[:d] == 'xx'

jwk = ::JWT:JWK.new(OpenSSL::PKey::RSA.new(2048), { kid: 'thekid', use:'sig' })
jwk.is_a?(::JWT::JWK::RSA)

jwk = ::JWT:JWK.new(OpenSSL::PKey::RSA.new(2048), { kid: 'thekid', use:'sig', n:'xxx'} ) # Would throw an error about mixing keypair and parameters is now allowed
````

Reply:
I like this idea :)
This is probably the most intuitive solution available for anyone using JWKs (especially the invocation of `new` with OpenSSL key and parameters).

However, the thing which may bite us later is any additional library specific parameters like `kid_generator`, which are no longer distinguishable from JWK parameters. Hypothetical scenario:

Someone wants to explicitly denote within a JWK the generation algorithm for the `kid` within a given ecosystem. Maybe there are at some point several such algorithms in standardized form and a JWK parameter `kid_generator` is being used to explicitly list the subset of algorithms being used in token headers.

Why would someone do this? No idea :)
But it would become a problem for us, as could any other parameter we might want to specify at this point (e.g. we were talking about validation options at one point).

We could pass two hashes to avoid the naming problem. An options hash and a parameters hash (in this order, to not break existing examples in the README):

```ruby
jwk = ::JWT:JWK.new(OpenSSL::PKey::RSA.new(2048),
                    { kid_generator: ::JWT::JWK::Thumbprint },
                    { kid: 'thekid', use:'sig' })
```
The rules are:
- Both the second and third parameter are optional
- The first parameter alone determines the key type and key, no matter whether it is given as a JWK hash or as an OpenSSL key
- The third parameter, if present, may not contain cryptographically relevant material
- For non-key stuff, the third parameter takes priority. E.g. if I specify `use` in both the first and third parameters, the latter will be used.
- The exception is `kid`, which may be present in either parameter for backwards compatibility. I would still argue that the priority should be `param_1 < param_3 < param_2`, but 2 and 3 may as well be the other way around

Reply:
I would like to see the more used parameters as the second/first parameter and the rest that control behaviour as the last one. Allowing `kid_generator` to be given in both for now. I don't see the kid_generator being part of a official RFC any time soon, you can freely quote me and make fun of me when that happens and we are in trouble:)

The keypair problem could be handled by memoizing the generated key until a jwk parameter changes.

Reply:
Swapped the parameters and implemented memoization. `kid` can no longer be specified in the `options` hash, where any future parameters for validation will be specified.
`kid_generator` can currently be specified in either the options or params hash.
I am going to merge this into the PR :)